### DISCUSSION TITLE
How to tile image watermark on an image by vips?

### DISCUSSION THREAD

**Discussion by darylpdavies** [libvips/libvips#2662] [Q&A] [answered]
I'm trying to convert this issue from a Text watermark to an Image watermark: https://github.com/libvips/libvips/issues/1237

Using this libvips example as a guide: https://github.com/libvips/ruby-vips/blob/d0f2c22d198415c724bef744f59c94882c632f70/example/watermark.rb

```
original_image = ImageProcessing::Vips.source(original)
watermark = "#{Rails.root}/public/icon-192.png"
watermark = watermark.rotate(-45)
watermarked = original_image.composite(watermark, mode: "over", gravity: "south-east", offset: [10, 10])
```

Just getting this far into the process, I'm getting this error:
`undefined method 'rotate' for nil:NilClass`

What can I do to convert my watermark into something vips can process?

**Comment by jcupitt** [comment, marked as answer]
Hi @daryld212, try something like:

```ruby
watermark = Vips::Image.new_from_file "some-file.png"
watermark = watermark.rotate(-45)
... etc.
```

**Reply by darylpdavies** [reply]
My reply got removed somehow, I'll try to rewrite it:

@jcupitt thanks, that definitely is now working - I have both the watermark image and my original image loading and ready to be put together.

I'm a little out of my depth to solve this error: `images do not have same numbers of bands`. I'm assuming `bandjoin` is the part that takes the alpha channel of the watermark and places it on the original?

Here is my updated script, after going through the ruby-vips docs I think I have most of it working - except the bandjoin line. I've also put in static dimensions (600x600) as I couldn't get the script to read the dimensions of the Tempfile.

```
    img = ImageProcessing::Vips.source(original)
    vips_img = img.call

    new_watermark = Vips::Image.new_from_file("#{Rails.root}/public/icon-192.png")
    new_watermark = new_watermark.rotate(-45)
    new_watermark = (new_watermark * 0.3).cast(:uchar)
    new_watermark = new_watermark.gravity(:centre, 200, 200)
    new_watermark = new_watermark.replicate(1 + 600 / 192, 1 + 600 / 192)
    new_watermark = new_watermark.crop(0, 0, 600, 600)

    overlay = new_watermark.new_from_image([255, 128, 128]).copy(interpretation: :srgb)
    overlay = overlay.bandjoin(new_watermark)

    new_watermarked = img.composite(overlay, mode: "over", gravity: "south-east", offset: [10, 10])
```

**Reply by jcupitt** [reply]
Hi again, composite tries to make everything into RGBA (four bands) before doing the composite operation. It will turn mono (one band) images into RGB, and turn RGB into RGBA, but it won't know what to do for stranger image types (RGBAA or RGBRGBA).

The main docs for `bandjoin` are here:

https://www.libvips.org/API/current/libvips-conversion.html#vips-bandjoin

So it joins two images band-wise. You can use a constant as well, eg. `image.bandjoin(255)` appends an extra band, setting it to the constant 255. You can `bandjoin([1, 2, 3])` to add three bands.

In your code you have:

```ruby
    overlay = new_watermark.new_from_image([255, 128, 128]).copy(interpretation: :srgb)
    overlay = overlay.bandjoin(new_watermark)
```

I suppose `new_watermark` is probably RGBA. The first line makes an RGB image the same size as `new_watermark` set to a reddish colour, the second line appends the original watermark RGBA, making an RGBARGB (seven band) image. When you later try to composite, it fails with the error.

Did you want to copy the alpha from `new_watermark`? You could try swapping that second line for:

```ruby
    overlay = overlay.bandjoin(new_watermark[3] * 0.3)
```

That'll get the alpha (band 3, numbering from 0), scale it down (ie. make it more transparent) and append it to your RGB.

I made you a tiny demo prog:

```ruby
#!/usr/bin/ruby

require 'vips'

if ARGV.length != 3
  puts "usage: ./watermark-image.rb IN OUT WATERMARK"
  exit
end

# open the background image in sequential mode: we can stream this and
# save a good chunk of memory and time
image = Vips::Image.new_from_file ARGV[0], :access => :sequential

# we need to reuse the overlay many times, so we can't stream it
watermark = Vips::Image.new_from_file ARGV[2]

# make sure we have an RGB image (it could be mono or CMYK, for example)
watermark = watermark.colourspace "srgb"

# the watermark will need an alpha channel
if ! watermark.has_alpha?
  watermark = watermark.add_alpha
end

# now we know we have RGBA, make the A semi-transparent and tilt it
watermark = (watermark * [1, 1, 1, 0.3]).rotate(-45)

# tile and crop to the same size as the 
# background image
overlay = watermark \
  .replicate(1 + image.width / watermark.width, 
             1 + image.height / watermark.height) \
  .crop(0, 0, image.width, image.height)

image = image.composite overlay, "over"

image.write_to_file ARGV[1]
```

I can run it like this:

```
$ ./watermark-image.rb ~/pics/nina.jpg x.jpg ~/pics/lion.svg
```

To make:

![x](https://user-images.githubusercontent.com/580843/154231073-6f63b925-1bea-4d41-8a01-5bb601ab6db0.jpg)

**Reply by jcupitt** [reply]
I deleted your other reply, sorry, there were two and I wasn't sure which one you wanted to keep.

It'd be good to paste the libvips C docs into the ruby-doc sources. We have all those docs as XML -- for example, if you build libvips from source you'll see a file called `libvips/Vips-8.0.gir` containing this:

```xml
    <function name="bandjoin" c:identifier="vips_bandjoin" introspectable="0">
      <doc xml:space="preserve"
           filename="conversion/bandjoin.c"
           line="206">Join a set of images together, bandwise.

If the images
have n and m bands, then the output image will have n + m
bands, with the first n coming from the first image and the last m
from the second.

If the images differ in size, the smaller images are enlarged to match the
larger by adding zero pixels along the bottom and right.

The input images are cast up to the smallest common type (see table
Smallest common format in
&lt;link linkend="libvips-arithmetic"&gt;arithmetic&lt;/link&gt;).

See also: vips_insert().</doc>
      <source-position filename="include/vips/conversion.h" line="232"/>
      <return-value transfer-ownership="none">
        <doc xml:space="preserve"
             filename="conversion/bandjoin.c"
             line="229">0 on success, -1 on error</doc>
        <type name="gint" c:type="int"/>
      </return-value>
      <parameters>
        <parameter name="in" transfer-ownership="none">
          <doc xml:space="preserve"
               filename="conversion/bandjoin.c"
               line="208">array of input images</doc>
          <array length="2" zero-terminated="0" c:type="VipsImage**">
            <type name="Image" c:type="VipsImage*"/>
          </array>
        </parameter>
        <parameter name="out"
                   direction="out"
                   caller-allocates="0"
                   transfer-ownership="full">
          <doc xml:space="preserve"
               filename="conversion/bandjoin.c"
               line="209">output image</doc>
          <type name="Image" c:type="VipsImage**"/>
        </parameter>
        <parameter name="n" transfer-ownership="none">
          <doc xml:space="preserve"
               filename="conversion/bandjoin.c"
               line="210">number of input images</doc>
          <type name="gint" c:type="int"/>
        </parameter>
        <parameter name="..." transfer-ownership="none">
          <doc xml:space="preserve"
               filename="conversion/bandjoin.c"
               line="211">%NULL-terminated list of optional named arguments</doc>
          <varargs/>
        </parameter>
      </parameters>
    </function>
```

It wouldn't be hard to reformat it for rdoc, but one one's found the time.
